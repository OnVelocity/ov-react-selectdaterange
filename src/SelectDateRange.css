.SelectDateRange {
	max-width: 350px;
}

.SelectDateRange .week {
	display: flex;
}

.SelectDateRange .day {
	flex: 0 0 auto;
	min-width: 50px;
	background-color: aliceblue;
	color: #000;
	margin: 0;
	border: 1px solid transparent;
	cursor: pointer;
}

.SelectDateRange .--is-not-in-start-month {
	background-color: #f5f5f5;
	color: #aaaaaa;
}

.SelectDateRange .--is-today {
	background-color: #cad8d6;
	color: #940d16;
}

.SelectDateRange .--is-selected-date {
	border: 1px solid darkred;
}
