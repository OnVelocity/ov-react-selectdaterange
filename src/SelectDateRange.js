import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
	getOffsetDate,
	getOffsetMonth,
	getOffsetYear,
	getStartOfMonthDate,
	getStartOfWeekDate,
	isDateInStartDateMonth,
	isDateSelectedDate,
	isDateToday
} from './date'
import './SelectDateRange.css'

export class SelectDateRange extends Component {
	static propTypes = {
		date: PropTypes.instanceOf(Date)
	}

	static defaultProps = {
		date: new Date(Date.now()),
		onDateSelected: date => console.log(`selected date ${date.toISOString()}`)
	}

	constructor(props) {
		super(props)
		this.state = {
			date: props.date,
			selectedDate: null
		}
	}

	render() {
		return (
			<div className="SelectDateRange">
				<div>
					<button onClick={() => this.previousYear()}>Previous Year</button>
					<span>{this.state.date.getFullYear()}</span>
					<button onClick={() => this.nextYear()}>Next Year</button>
				</div>
				<div>
					<button onClick={() => this.previousMonth()}>Previous Month</button>
					<MonthName date={this.state.date} />
					<button onClick={() => this.nextMonth()}>Next Month</button>
				</div>
				<div className="week">
					<div className="day">Sun</div>
					<div className="day">Mon</div>
					<div className="day">Tue</div>
					<div className="day">Wed</div>
					<div className="day">Thu</div>
					<div className="day">Fri</div>
					<div className="day">Sat</div>
				</div>
				<div className="calendar">
					<CalendarMonth
						startDate={this.state.date}
						selectedDate={this.state.selectedDate}
						onDateSelected={date => this.selectDate(date)}
					/>
				</div>
			</div>
		)
	}

	previousYear() {
		const date = getOffsetYear(this.state.date, -1)
		this.setState({ date })
	}

	nextYear() {
		const date = getOffsetYear(this.state.date, 1)
		this.setState({ date })
	}

	previousMonth() {
		const date = getOffsetMonth(this.state.date, -1)
		this.setState({ date })
	}

	nextMonth() {
		const date = getOffsetMonth(this.state.date, 1)
		this.setState({ date })
	}

	selectDate(date) {
		this.setState({ selectedDate: date })
		this.props.onDateSelected(date)
	}
}

export function CalendarMonth({ startDate, selectedDate, onDateSelected }) {
	const monthStart = getStartOfMonthDate(startDate)
	const weekStart = getStartOfWeekDate(monthStart)
	const weeks = []
	for (let i = 0; i < 6; i++) {
		const offset = i * 7
		const date = getOffsetDate(weekStart, offset)
		weeks.push(
			<CalendarWeek
				key={date}
				startOfWeekDate={date}
				startDate={startDate}
				selectedDate={selectedDate}
				onDateSelected={onDateSelected}
			/>
		)
	}
	return weeks
}

export function CalendarWeek({
	startOfWeekDate,
	startDate,
	selectedDate,
	onDateSelected
}) {
	const week = []
	for (let i = 0; i < 7; i++) {
		const date = getOffsetDate(startOfWeekDate, i)
		week.push(
			<CalendarDay
				key={date}
				date={date}
				startDate={startDate}
				selectedDate={selectedDate}
				onDateSelected={onDateSelected}
			/>
		)
	}
	return (
		<div key={startOfWeekDate} className="week">
			{week}
		</div>
	)
}

export function CalendarDay({ date, startDate, selectedDate = null, onDateSelected }) {
	const day = date.getDate()
	const isToday = isDateToday(date)
	const inMonth = isDateInStartDateMonth(date, startDate)
	const classNames = {
		day: true,
		'--is-not-in-start-month': !inMonth,
		'--is-today': isToday
	}
	if (selectedDate instanceof Date) {
		classNames['--is-selected-date'] = isDateSelectedDate(date, selectedDate)
	}
	const className = Object.keys(classNames)
		.reduce((classes, name) => {
			if (classNames[name]) {
				classes.push(name)
			}
			return classes
		}, [])
		.join(' ')
	return (
		<div
			key={day}
			className={className}
			onClick={() => {
				onDateSelected(date)
			}}
		>
			{day}
		</div>
	)
}

export function MonthName({ date }) {
	const month = date && date.getMonth()
	switch (month) {
		case 0:
			return <span>January</span>
		case 1:
			return <span>February</span>
		case 2:
			return <span>March</span>
		case 3:
			return <span>April</span>
		case 4:
			return <span>May</span>
		case 5:
			return <span>June</span>
		case 6:
			return <span>July</span>
		case 7:
			return <span>August</span>
		case 8:
			return <span>September</span>
		case 9:
			return <span>October</span>
		case 10:
			return <span>November</span>
		case 11:
			return <span>December</span>
		default:
			return <span> </span>
	}
}
