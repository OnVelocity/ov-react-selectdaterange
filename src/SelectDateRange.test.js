/*global describe, it, expect, jest, beforeEach */
import React from 'react'
import ReactDOM from 'react-dom'
import ReactDomTestUtils from 'react-dom/test-utils'
import renderer from 'react-test-renderer'
import {SelectDateRange} from "./SelectDateRange";

describe('SelectDateRange', () => {
	let props
	beforeEach(() => {
		props = {
			date: new Date(2017, 9, 14, 0, 0, 0),
			selectedDate: new Date(2017, 9, 12, 0, 0, 0),
			onDateSelected: jest.fn()
		}
	})
	it('renders', () => {
		expect.assertions(1)
		const tree = renderer.create(
			<SelectDateRange
				date={props.date}
				selectedDate={props.selectedDate}
				onDateSelected={props.onDateSelected}
			/>
		)
		expect(tree.toJSON()).toMatchSnapshot()
	})
	it('calls default onDateSelected prop', () => {
		expect.assertions(1)
		const spy = jest.spyOn(console, 'log')
		const div = document.createElement('div')
		ReactDOM.render(
			<SelectDateRange
				date={props.date}
				selectedDate={props.selectedDate}
			/>,
			div
		)
		const day = div.querySelectorAll('.day')[10]
		ReactDomTestUtils.Simulate.click(day, {})
		expect(spy).toBeCalledWith('selected date 2017-10-04T04:00:00.000Z')
		spy.mockRestore()
	})
	it('calls onDateSelected callback', () => {
		expect.assertions(1)
		const div = document.createElement('div')
		ReactDOM.render(<SelectDateRange {...props}/>, div)
		const day = div.querySelectorAll('.day')[10]
		ReactDomTestUtils.Simulate.click(day, {})
		expect(props.onDateSelected).toHaveBeenCalledWith(new Date(2017, 9, 4))
	})
	it('calls previousYear()', () => {
		expect.assertions(1)
		const div = document.createElement('div')
		ReactDOM.render(<SelectDateRange {...props}/>, div)
		const button = div.querySelectorAll('button')[0]
		ReactDomTestUtils.Simulate.click(button, {})
		expect(div.innerHTML).toMatchSnapshot()
	})
	it('calls nextYear()', () => {
		expect.assertions(1)
		const div = document.createElement('div')
		ReactDOM.render(<SelectDateRange {...props}/>, div)
		const button = div.querySelectorAll('button')[1]
		ReactDomTestUtils.Simulate.click(button, {})
		expect(div.innerHTML).toMatchSnapshot()
	})
	it('calls previousMonth()', () => {
		expect.assertions(1)
		const div = document.createElement('div')
		ReactDOM.render(<SelectDateRange {...props}/>, div)
		const button = div.querySelectorAll('button')[2]
		ReactDomTestUtils.Simulate.click(button, {})
		expect(div.innerHTML).toMatchSnapshot()
	})
	it('calls nextMonth()', () => {
		expect.assertions(1)
		const div = document.createElement('div')
		ReactDOM.render(<SelectDateRange {...props}/>, div)
		const button = div.querySelectorAll('button')[3]
		ReactDomTestUtils.Simulate.click(button, {})
		expect(div.innerHTML).toMatchSnapshot()
	})
})
