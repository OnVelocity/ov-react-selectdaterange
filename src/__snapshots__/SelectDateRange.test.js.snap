// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectDateRange calls nextMonth() 1`] = `"<div class=\\"SelectDateRange\\"><div><button>Previous Year</button><span>2017</span><button>Next Year</button></div><div><button>Previous Month</button><span>November</span><button>Next Month</button></div><div class=\\"week\\"><div class=\\"day\\">Sun</div><div class=\\"day\\">Mon</div><div class=\\"day\\">Tue</div><div class=\\"day\\">Wed</div><div class=\\"day\\">Thu</div><div class=\\"day\\">Fri</div><div class=\\"day\\">Sat</div></div><div class=\\"calendar\\"><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">29</div><div class=\\"day --is-not-in-start-month\\">30</div><div class=\\"day --is-not-in-start-month\\">31</div><div class=\\"day\\">1</div><div class=\\"day\\">2</div><div class=\\"day\\">3</div><div class=\\"day\\">4</div></div><div class=\\"week\\"><div class=\\"day\\">5</div><div class=\\"day\\">6</div><div class=\\"day\\">7</div><div class=\\"day\\">8</div><div class=\\"day\\">9</div><div class=\\"day\\">10</div><div class=\\"day\\">11</div></div><div class=\\"week\\"><div class=\\"day\\">12</div><div class=\\"day\\">13</div><div class=\\"day\\">14</div><div class=\\"day\\">15</div><div class=\\"day\\">16</div><div class=\\"day\\">17</div><div class=\\"day\\">18</div></div><div class=\\"week\\"><div class=\\"day\\">19</div><div class=\\"day\\">20</div><div class=\\"day\\">21</div><div class=\\"day\\">22</div><div class=\\"day\\">23</div><div class=\\"day\\">24</div><div class=\\"day\\">25</div></div><div class=\\"week\\"><div class=\\"day\\">26</div><div class=\\"day\\">27</div><div class=\\"day\\">28</div><div class=\\"day\\">29</div><div class=\\"day\\">30</div><div class=\\"day --is-not-in-start-month\\">1</div><div class=\\"day --is-not-in-start-month\\">2</div></div><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">3</div><div class=\\"day --is-not-in-start-month\\">4</div><div class=\\"day --is-not-in-start-month\\">5</div><div class=\\"day --is-not-in-start-month\\">6</div><div class=\\"day --is-not-in-start-month\\">7</div><div class=\\"day --is-not-in-start-month\\">8</div><div class=\\"day --is-not-in-start-month\\">9</div></div></div></div>"`;

exports[`SelectDateRange calls nextYear() 1`] = `"<div class=\\"SelectDateRange\\"><div><button>Previous Year</button><span>2018</span><button>Next Year</button></div><div><button>Previous Month</button><span>October</span><button>Next Month</button></div><div class=\\"week\\"><div class=\\"day\\">Sun</div><div class=\\"day\\">Mon</div><div class=\\"day\\">Tue</div><div class=\\"day\\">Wed</div><div class=\\"day\\">Thu</div><div class=\\"day\\">Fri</div><div class=\\"day\\">Sat</div></div><div class=\\"calendar\\"><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">30</div><div class=\\"day\\">1</div><div class=\\"day\\">2</div><div class=\\"day\\">3</div><div class=\\"day\\">4</div><div class=\\"day\\">5</div><div class=\\"day\\">6</div></div><div class=\\"week\\"><div class=\\"day\\">7</div><div class=\\"day\\">8</div><div class=\\"day\\">9</div><div class=\\"day\\">10</div><div class=\\"day\\">11</div><div class=\\"day\\">12</div><div class=\\"day\\">13</div></div><div class=\\"week\\"><div class=\\"day\\">14</div><div class=\\"day\\">15</div><div class=\\"day\\">16</div><div class=\\"day\\">17</div><div class=\\"day\\">18</div><div class=\\"day\\">19</div><div class=\\"day\\">20</div></div><div class=\\"week\\"><div class=\\"day\\">21</div><div class=\\"day\\">22</div><div class=\\"day\\">23</div><div class=\\"day\\">24</div><div class=\\"day\\">25</div><div class=\\"day\\">26</div><div class=\\"day\\">27</div></div><div class=\\"week\\"><div class=\\"day\\">28</div><div class=\\"day\\">29</div><div class=\\"day\\">30</div><div class=\\"day\\">31</div><div class=\\"day --is-not-in-start-month\\">1</div><div class=\\"day --is-not-in-start-month\\">2</div><div class=\\"day --is-not-in-start-month\\">3</div></div><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">4</div><div class=\\"day --is-not-in-start-month\\">5</div><div class=\\"day --is-not-in-start-month\\">6</div><div class=\\"day --is-not-in-start-month\\">7</div><div class=\\"day --is-not-in-start-month\\">8</div><div class=\\"day --is-not-in-start-month\\">9</div><div class=\\"day --is-not-in-start-month\\">10</div></div></div></div>"`;

exports[`SelectDateRange calls previousMonth() 1`] = `"<div class=\\"SelectDateRange\\"><div><button>Previous Year</button><span>2017</span><button>Next Year</button></div><div><button>Previous Month</button><span>September</span><button>Next Month</button></div><div class=\\"week\\"><div class=\\"day\\">Sun</div><div class=\\"day\\">Mon</div><div class=\\"day\\">Tue</div><div class=\\"day\\">Wed</div><div class=\\"day\\">Thu</div><div class=\\"day\\">Fri</div><div class=\\"day\\">Sat</div></div><div class=\\"calendar\\"><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">27</div><div class=\\"day --is-not-in-start-month\\">28</div><div class=\\"day --is-not-in-start-month\\">29</div><div class=\\"day --is-not-in-start-month\\">30</div><div class=\\"day --is-not-in-start-month\\">31</div><div class=\\"day\\">1</div><div class=\\"day\\">2</div></div><div class=\\"week\\"><div class=\\"day\\">3</div><div class=\\"day\\">4</div><div class=\\"day\\">5</div><div class=\\"day\\">6</div><div class=\\"day\\">7</div><div class=\\"day\\">8</div><div class=\\"day\\">9</div></div><div class=\\"week\\"><div class=\\"day\\">10</div><div class=\\"day\\">11</div><div class=\\"day\\">12</div><div class=\\"day\\">13</div><div class=\\"day\\">14</div><div class=\\"day\\">15</div><div class=\\"day\\">16</div></div><div class=\\"week\\"><div class=\\"day\\">17</div><div class=\\"day\\">18</div><div class=\\"day\\">19</div><div class=\\"day\\">20</div><div class=\\"day\\">21</div><div class=\\"day\\">22</div><div class=\\"day\\">23</div></div><div class=\\"week\\"><div class=\\"day\\">24</div><div class=\\"day\\">25</div><div class=\\"day\\">26</div><div class=\\"day\\">27</div><div class=\\"day\\">28</div><div class=\\"day\\">29</div><div class=\\"day\\">30</div></div><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">1</div><div class=\\"day --is-not-in-start-month\\">2</div><div class=\\"day --is-not-in-start-month\\">3</div><div class=\\"day --is-not-in-start-month\\">4</div><div class=\\"day --is-not-in-start-month\\">5</div><div class=\\"day --is-not-in-start-month\\">6</div><div class=\\"day --is-not-in-start-month\\">7</div></div></div></div>"`;

exports[`SelectDateRange calls previousYear() 1`] = `"<div class=\\"SelectDateRange\\"><div><button>Previous Year</button><span>2016</span><button>Next Year</button></div><div><button>Previous Month</button><span>October</span><button>Next Month</button></div><div class=\\"week\\"><div class=\\"day\\">Sun</div><div class=\\"day\\">Mon</div><div class=\\"day\\">Tue</div><div class=\\"day\\">Wed</div><div class=\\"day\\">Thu</div><div class=\\"day\\">Fri</div><div class=\\"day\\">Sat</div></div><div class=\\"calendar\\"><div class=\\"week\\"><div class=\\"day --is-not-in-start-month\\">25</div><div class=\\"day --is-not-in-start-month\\">26</div><div class=\\"day --is-not-in-start-month\\">27</div><div class=\\"day --is-not-in-start-month\\">28</div><div class=\\"day --is-not-in-start-month\\">29</div><div class=\\"day --is-not-in-start-month\\">30</div><div class=\\"day\\">1</div></div><div class=\\"week\\"><div class=\\"day\\">2</div><div class=\\"day\\">3</div><div class=\\"day\\">4</div><div class=\\"day\\">5</div><div class=\\"day\\">6</div><div class=\\"day\\">7</div><div class=\\"day\\">8</div></div><div class=\\"week\\"><div class=\\"day\\">9</div><div class=\\"day\\">10</div><div class=\\"day\\">11</div><div class=\\"day\\">12</div><div class=\\"day\\">13</div><div class=\\"day\\">14</div><div class=\\"day\\">15</div></div><div class=\\"week\\"><div class=\\"day\\">16</div><div class=\\"day\\">17</div><div class=\\"day\\">18</div><div class=\\"day\\">19</div><div class=\\"day\\">20</div><div class=\\"day\\">21</div><div class=\\"day\\">22</div></div><div class=\\"week\\"><div class=\\"day\\">23</div><div class=\\"day\\">24</div><div class=\\"day\\">25</div><div class=\\"day\\">26</div><div class=\\"day\\">27</div><div class=\\"day\\">28</div><div class=\\"day\\">29</div></div><div class=\\"week\\"><div class=\\"day\\">30</div><div class=\\"day\\">31</div><div class=\\"day --is-not-in-start-month\\">1</div><div class=\\"day --is-not-in-start-month\\">2</div><div class=\\"day --is-not-in-start-month\\">3</div><div class=\\"day --is-not-in-start-month\\">4</div><div class=\\"day --is-not-in-start-month\\">5</div></div></div></div>"`;

exports[`SelectDateRange renders 1`] = `
<div
  className="SelectDateRange"
>
  <div>
    <button
      onClick={[Function]}
    >
      Previous Year
    </button>
    <span>
      2017
    </span>
    <button
      onClick={[Function]}
    >
      Next Year
    </button>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Previous Month
    </button>
    <span>
      October
    </span>
    <button
      onClick={[Function]}
    >
      Next Month
    </button>
  </div>
  <div
    className="week"
  >
    <div
      className="day"
    >
      Sun
    </div>
    <div
      className="day"
    >
      Mon
    </div>
    <div
      className="day"
    >
      Tue
    </div>
    <div
      className="day"
    >
      Wed
    </div>
    <div
      className="day"
    >
      Thu
    </div>
    <div
      className="day"
    >
      Fri
    </div>
    <div
      className="day"
    >
      Sat
    </div>
  </div>
  <div
    className="calendar"
  >
    <div
      className="week"
    >
      <div
        className="day"
        onClick={[Function]}
      >
        1
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        2
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        3
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        4
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        5
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        6
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        7
      </div>
    </div>
    <div
      className="week"
    >
      <div
        className="day"
        onClick={[Function]}
      >
        8
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        9
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        10
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        11
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        12
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        13
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        14
      </div>
    </div>
    <div
      className="week"
    >
      <div
        className="day"
        onClick={[Function]}
      >
        15
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        16
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        17
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        18
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        19
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        20
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        21
      </div>
    </div>
    <div
      className="week"
    >
      <div
        className="day"
        onClick={[Function]}
      >
        22
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        23
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        24
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        25
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        26
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        27
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        28
      </div>
    </div>
    <div
      className="week"
    >
      <div
        className="day"
        onClick={[Function]}
      >
        29
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        30
      </div>
      <div
        className="day"
        onClick={[Function]}
      >
        31
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        1
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        2
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        3
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        4
      </div>
    </div>
    <div
      className="week"
    >
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        5
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        6
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        7
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        8
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        9
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        10
      </div>
      <div
        className="day --is-not-in-start-month"
        onClick={[Function]}
      >
        11
      </div>
    </div>
  </div>
</div>
`;
